x/emissions/keeper/msgserver/msg_server_whitelist.go

```diff
@@ -40,18 +43,334 @@ func (ms msgServer) RemoveFromWhitelistAdmin(ctx context.Context, msg *types.Rem
+	}
+
+	// Validate the address
+	if err := ms.k.ValidateStringIsBech32(msg.Address); err != nil {
+		return nil, err
+	}
+
+	// Remove the address from the whitelist
+	return &types.RemoveFromWhitelistAdminResponse{}, ms.k.RemoveWhitelistAdmin(ctx, msg.Address)
+}
+
+func (ms msgServer) AddToGlobalWhitelist(ctx context.Context, msg *types.AddToGlobalWhitelistRequest) (_ *types.AddToGlobalWhitelistResponse, err error) {
+	defer metrics.RecordMetrics("AddToGlobalWhitelist", time.Now(), &err)
+
+	// Validate the sender address
+	err = ms.k.ValidateStringIsBech32(msg.Sender)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check that sender is permitted to update global whitelists
+	canUpdate, err := ms.k.CanUpdateGlobalWhitelists(ctx, msg.Sender)
+	if err != nil {
+		return nil, err
+	} else if !canUpdate {
+		return nil, types.ErrNotPermittedToUpdateGlobalWhitelist
+	}
+
+	// Validate the address
+	if err := ms.k.ValidateStringIsBech32(msg.Address); err != nil {
+		return nil, err
+	}
+
+	// Add the address to the whitelist
+	return &types.AddToGlobalWhitelistResponse{}, ms.k.AddToGlobalWhitelist(ctx, msg.Address)
+}
+
+func (ms msgServer) RemoveFromGlobalWhitelist(ctx context.Context, msg *types.RemoveFromGlobalWhitelistRequest) (_ *types.RemoveFromGlobalWhitelistResponse, err error) {
+	defer metrics.RecordMetrics("RemoveFromGlobalWhitelist", time.Now(), &err)
+
+	// Validate the sender address
+	err = ms.k.ValidateStringIsBech32(msg.Sender)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check that sender is permitted to update global whitelists
+	canUpdate, err := ms.k.CanUpdateGlobalWhitelists(ctx, msg.Sender)
 	if err != nil {
 		return nil, err
-	} else if !isAdmin {
-		return nil, types.ErrNotWhitelistAdmin
+	} else if !canUpdate {
+		return nil, types.ErrNotPermittedToUpdateGlobalWhitelist
 	}
+
 	// Validate the address
 	if err := ms.k.ValidateStringIsBech32(msg.Address); err != nil {
 		return nil, err
 	}
+
 	// Remove the address from the whitelist
-	err = ms.k.RemoveWhitelistAdmin(ctx, msg.Address)
-	return &types.RemoveFromWhitelistAdminResponse{}, err
+	return &types.RemoveFromGlobalWhitelistResponse{}, ms.k.RemoveFromGlobalWhitelist(ctx, msg.Address)
+}
+
+func (ms msgServer) EnableTopicWorkerWhitelist(ctx context.Context, msg *types.EnableTopicWorkerWhitelistRequest) (_ *types.EnableTopicWorkerWhitelistResponse, err error) {
+	defer metrics.RecordMetrics("EnableTopicWorkerWhitelist", time.Now(), &err)
+
+	exists, err := ms.k.TopicExists(ctx, msg.TopicId)
+	if err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, errorsmod.Wrap(types.ErrNotFound, "topic does not exist")
+	}
+
+	// Validate the address
+	if err := ms.k.ValidateStringIsBech32(msg.Sender); err != nil {
+		return nil, err
+	}
+	canUpdate, err := ms.k.CanUpdateTopicWhitelist(ctx, msg.TopicId, msg.Sender)
+	if err != nil {
+		return nil, err
+	} else if !canUpdate {
+		return nil, types.ErrNotPermittedToUpdateTopicWhitelist
+	}
```

Comment:
Here, you're checking if the topic exists and then querying it again inside `CanUpdateTopicWhitelist.` You can just do one query to check both topic's existence, then topic creator. Please do the same thing in the others.

Reply:
Just have to be sure to catch a particular "not found" error first but then should avoid an extra query, yes.